pdns/opensslsigners.cc

```diff
@@ -5,9 +5,306 @@
+  }
+
+  return string((char*) signature, signatureLen);
+}
+
+
+bool OpenSSLRSADNSCryptoKeyEngine::verify(const std::string& msg, const std::string& signature) const
+{
+  string hash = this->hash(msg);
+  int hashKind;
+
+  if (hash.size() == SHA_DIGEST_LENGTH) {
+    hashKind = NID_sha1;
+  }
+  else if (hash.size() == SHA256_DIGEST_LENGTH) {
+    hashKind = NID_sha256;
+  }
+  else {
+    hashKind = NID_sha512;
+  }
+
+  int ret = RSA_verify(hashKind, (const unsigned char*) hash.c_str(), hash.length(), (unsigned char*) signature.c_str(), signature.length(), d_key);
+
+  return (ret == 1);
+}
+
+
+std::string OpenSSLRSADNSCryptoKeyEngine::getPubKeyHash() const
+{
+  unsigned char tmp[std::max(BN_num_bytes(d_key->e), BN_num_bytes(d_key->n))];
+  unsigned char hash[SHA_DIGEST_LENGTH];
+  SHA_CTX ctx;
+
+  int res = SHA1_Init(&ctx);
+
+  if (res != 1) {
+    throw runtime_error(getName()+" failed to init hash context for generating the public key hash");
+  }
+
+  int len = BN_bn2bin(d_key->e, tmp);
+  res = SHA1_Update(&ctx, tmp, len);
+  if (res != 1) {
+    throw runtime_error(getName()+" failed to update hash context for generating the public key hash");
+  }
+
+  len = BN_bn2bin(d_key->n, tmp);
+  res = SHA1_Update(&ctx, tmp, len);
+  if (res != 1) {
+    throw runtime_error(getName()+" failed to update hash context for generating the public key hash");
+  }
+
+  res = SHA1_Final(hash, &ctx);
+  if (res != 1) {
+    throw runtime_error(getName()+" failed to finish hash context for generating the public key hash");
+  }
+
+  return string((char*) hash, sizeof(hash));
+}
+
+
+std::string OpenSSLRSADNSCryptoKeyEngine::getPublicKeyString() const
+{
+  string keystring;
+  unsigned char tmp[std::max(BN_num_bytes(d_key->e), BN_num_bytes(d_key->n))];
+
+  int len = BN_bn2bin(d_key->e, tmp);
+  if (len < 255) {
+    keystring.assign(1, (char) (unsigned int) len);
+  } else {
+    keystring.assign(1, 0);
+    uint16_t tempLen = len;
+    tempLen = htons(tempLen);
+    keystring.append((char*)&tempLen, 2);
+  }
+  keystring.append((char *) tmp, len);
+
+  len = BN_bn2bin(d_key->n, tmp);
+  keystring.append((char *) tmp, len);
+
+  return keystring;
+}
+
+
+void OpenSSLRSADNSCryptoKeyEngine::fromISCMap(DNSKEYRecordContent& drc, std::map<std::string, std::string>& stormap)
+{
+  string sline;
+  string key,value;
+  typedef map<string, BIGNUM**> places_t;
+  places_t places;
+
+  places["Modulus"]=&d_key->n;
+  places["PublicExponent"]=&d_key->e;
+  places["PrivateExponent"]=&d_key->d;
+  places["Prime1"]=&d_key->p;
+  places["Prime2"]=&d_key->q;
+  places["Exponent1"]=&d_key->dmp1;
+  places["Exponent2"]=&d_key->dmq1;
+  places["Coefficient"]=&d_key->iqmp;
+
+  drc.d_algorithm = pdns_stou(stormap["algorithm"]);
+
+  string raw;
+  for(const places_t::value_type& val :  places) {
+    raw=stormap[toLower(val.first)];
+    *val.second = BN_bin2bn((unsigned char*) raw.c_str(), raw.length(), NULL);
```

Comment:
it is possible to reuse the existing  *val.second structure (s/NULL/*val.second)
Maybe we even leak memory here.